Enable per-token classification in RoBERTa

prihoda opened the pull request:
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/pytorch/fairseq/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?   
- [x] Did you write any new necessary tests?  

## What does this PR do?
Fixes https://github.com/pytorch/fairseq/issues/1708

prihoda commented on fairseq/models/roberta/model.py:
```diff
@@ -230,15 +231,18 @@ def forward(self, features, masked_tokens=None, **kwargs):
+            x = x[:, 0, :]  # take <s> token (equiv. to [CLS])
         x = self.dropout(x)
         x = self.dense(x)
         x = self.activation_fn(x)
```
@myleott Would it make sense to add `layer_norm` here just like in the `RobertaLMHead`? https://github.com/pytorch/fairseq/pull/1709/files#diff-ad4f875b91a61dcc1511c93c8e38cc4bL224

prihoda commented on fairseq/tasks/sequence_tagging.py:
```diff
@@ -0,0 +1,242 @@
+            label_dataset = make_dataset('label', self.label_dictionary)
+            if label_dataset is not None:
+                dataset.update(
+                    target=RightPadDataset( # append -1 as padding token
```
@myleott I'm removing the `eos` token and padding the dataset with `-1` to keep the class dictionary in the original range (`0,1,2,...`) and to enable masking. Is there a better way?

myleott replied:
This seems fine, but please add a comment in the code

prihoda commented on fairseq/tasks/sequence_tagging.py:
```diff
@@ -0,0 +1,242 @@
+                    )
+                )
+        else:
+            raise NotImplementedError()
```
@myleott Can I keep this here or shall I rather remove the regression option before it's implemented?

prihoda commented on fairseq/criterions/sequence_tagging.py:
```diff
@@ -0,0 +1,101 @@
+
+        if not self.args.regression_target:
+            logits = logits.view(-1, logits.size(-1))
+            loss = F.nll_loss(
```
@myleott Can you please validate this loss calculation?

myleott replied:
Looks good, but it'd be great if you could test this on a real sequence tagging task first.

prihoda replied:
I added an example with a README 😉 https://github.com/pytorch/fairseq/pull/1709/files#diff-dcc107dc4ee613a610d8174151f0afc4

This made me realize that it might be tricky to use sequence tagging with BPE, since you will have to provide a label for each input BPE token. I clarified that in the README, hope that is enough.

I also added `predict_class` and `predict_tags` convenience methods and used them in the classification and tagging READMEs.

prihoda replied:
By the way, shouldn't I name this task `sentence_tagging` instead of `sequence_tagging`, since we already have `sentence_prediction` and `sentence_ranking`?

prihoda commented:
@myleott What else is needed to make this mergeable? Do I need to create a dedicated RoBERTa example README for sequence tagging?

myleott commented on fairseq/criterions/sequence_tagging.py:
```diff
@@ -0,0 +1,101 @@
+    def add_args(parser):
+        # fmt: off
+        parser.add_argument('--classification-head-name',
+                            default='sentence_classification_head',
```
`sequence_tagging_head`

myleott commented on fairseq/tasks/sequence_tagging.py:
```diff
@@ -0,0 +1,242 @@
+            src_tokens = input0
+        else:
+            if self.args.separator_token is not None:
+                input1 = PrependTokenDataset(input1, self.args.separator_token)
```
Do we ever have two inputs for this kind of task?

prihoda replied:
Good point, I removed this functionality

myleott commented on fairseq/tasks/sequence_tagging.py:
```diff
@@ -0,0 +1,242 @@
+        parser.add_argument('--truncate-sequence', action='store_true', default=False,
+                            help='truncate sequence to max-positions')
+        parser.add_argument('--add-prev-output-tokens', action='store_true', default=False,
+                            help='add prev_output_tokens to sample, used for encoder-decoder arch')
```
We can probably remove `--separator-token` and `--add-prev-output-tokens`

myleott commented on fairseq/tasks/sequence_tagging.py:
```diff
@@ -0,0 +1,242 @@
+                prev_output_tokens=prev_tokens_dataset,
+            )
+
+        if not self.args.regression_target:
```
Since we don't support add_prev_output_tokens or regression_target, shall we fold this into the `dataset` dictionary definition above?

myleott commented on fairseq/tasks/sequence_tagging.py:
```diff
@@ -0,0 +1,242 @@
+            )
+
+        if not self.args.regression_target:
+            label_dataset = make_dataset('label', self.label_dictionary)
```
We should add some kind of check that the lengths of label dataset match the length of the src_tokens dataset.

How about something like this:
```
dataset = {
    'id': IdDataset(),
    ...
    'target': label_dataset,
    '_assert_lengths_match': AssertSameLengthDataset(src_tokens, label_dataset),
}
```

AssertSameLengthDataset can just return `1` if the lengths of each item match, otherwise raise an exception.

prihoda commented on fairseq/models/roberta/model.py:
```diff
@@ -112,7 +112,7 @@ def forward(self, src_tokens, features_only=False, return_all_hiddens=False, cla
         return x, extra
 
-    def register_classification_head(self, name, num_classes=None, inner_dim=None, **kwargs):
+    def register_classification_head(self, name, num_classes=None, sequence_tagging=False, inner_dim=None, **kwargs):
```
And these would be stored in `self.classification_heads` or in separate `self.sequence_tagging_heads`?

prihoda commented on fairseq/criterions/sequence_tagging.py:
```diff
@@ -0,0 +1,101 @@
+        assert (
+            hasattr(model, 'classification_heads')
+            and self.args.classification_head_name in model.classification_heads
+        ), 'model must provide sentence classification head for --criterion=sequence_tagging'
```
@myleott How shall I handle this assertion when we are performing sequence tagging? Shall I check that either is present?

prihoda commented:
@myleott I added an example with a README 😉 https://github.com/pytorch/fairseq/pull/1709/files#diff-dcc107dc4ee613a610d8174151f0afc4

This made me realize that it might be tricky to use sequence tagging with BPE, since you will have to provide a label for each input BPE token. I clarified that in the README, hope that is enough.

I also added `predict_class` and `predict_tags` convenience methods and used them in the classification and tagging READMEs.

By the way, shouldn't I name this task `sentence_tagging` instead of `sequence_tagging`, since we already have `sentence_prediction` and `sentence_ranking`?

prihoda commented:
@myleott I updated the code to reflect #1729

prihoda commented:
@myleott I rebased the PR on develop. You mentioned that you want to add a separate `register_sequence_tagging_head` function, to make it clearer what models support sequence tagging vs. sentence-level classification. Should these be stored in `self.classification_heads` or in separate `self.sequence_tagging_heads`?

avi-jit commented:
@prihoda I'm trying out your branch feature/classify_per_token on a custom sequence tagging task, but am running into an error: `has_marked_unused_parameters_ INTERNAL ASSERT FAILED at /pytorch/torch/csrc/distributed/c10d/reducer.cpp:327` which resembles the issues pytorch/pytorch#32490 and pytorch/pytorch#31035.

Any help would be appreciated in debugging. I didn't want to paste the lengthy log here unless it's relevant to this issue. Thanks in advance :)

prihoda commented:
Hi @avi-jit, let me know if you find out how this could be related to this branch, at this moment I'm not sure how to find the cause of your issue. Can you first try training using a simpler `sentence_prediction` task?

avi-jit commented:
sure, i'll try that @prihoda. Could you confirm if feature/classify_per_token is the correct branch to be working with?

prihoda commented:
@avi-jit yes that's the one

apyspo commented:
Could you please also share preprocess_tagging.sh mentioned in the README file

prihoda commented:
Hi @apyspo I tried to look for this file but can't seem to find it, you would probably need to write the script yourself. It transforms the downloaded files into separate input and label files, one sentence per line.

apyspo commented:
> 
Thank you @prihoda , I have written a script and transformed into one sentence per line similar to top ten sample lines in the README file.

prihoda commented:
Hi @apyspo will you try to finish this pull request? We could ask the main contributors what's needed to have this merged.
